File: bin/strap-runner/session.js
import fs from "node:fs";
import path from "node:path";
import { strapWorkRoot } from "./env.js";

export function resolveSession(value, die) {
  const direct = path.resolve(value);
  if (fs.existsSync(direct) && fs.statSync(direct).isDirectory()) return direct;

  const sessions = path.join(strapWorkRoot(), "sessions");
  const candidate = path.join(sessions, value);
  if (fs.existsSync(candidate) && fs.statSync(candidate).isDirectory()) return candidate;

  const matches = fs.existsSync(sessions)
    ? fs.readdirSync(sessions, { withFileTypes: true })
      .filter((entry) => entry.isDirectory() && entry.name.includes(value))
      .map((entry) => path.join(sessions, entry.name))
    : [];

  if (matches.length === 1) return matches[0];
  if (matches.length > 1) die(`Ambiguous session: ${value}`);
  die(`Session not found: ${value}`);
}

File: subprojects/strap-main/src/strap_main/session.clj
(ns strap-main.session
  (:require [babashka.fs :as fs]
            [clojure.string :as str]
            [strap-main.env :as env]))

(defn resolve-session [value die]
  (let [direct (fs/absolutize value)
        sessions (fs/path (env/strap-work-root) "sessions")
        candidate (fs/path sessions value)
        matches (if (fs/directory? sessions)
                  (filter #(and (fs/directory? %)
                                (str/includes? (fs/file-name %) value))
                          (fs/list-dir sessions))
                  [])]
    (cond
      (fs/directory? direct) (str direct)
      (fs/directory? candidate) (str candidate)
      (= 1 (count matches)) (str (first matches))
      (< 1 (count matches)) (die (str "Ambiguous session: " value))
      :else (die (str "Session not found: " value)))))
